src/services/suggestionDiagnostics.ts

```diff
@@ -17,12 +17,30 @@ namespace ts {
             switch (node.kind) {
                 case SyntaxKind.FunctionDeclaration:
                 case SyntaxKind.FunctionExpression:
+
                     if (isJsFile) {
                         const symbol = node.symbol;
                         if (symbol.members && (symbol.members.size > 0)) {
                             diags.push(createDiagnosticForNode(isVariableDeclaration(node.parent) ? node.parent.name : node, Diagnostics.This_constructor_function_may_be_converted_to_a_class_declaration));
                         }
                     }
+                    if (node.modifiers && isAsyncFunction(node)) {
+                        break;
+                    }
+
+                    const returnType = checker.getReturnTypeOfSignature(checker.getSignatureFromDeclaration(<FunctionDeclaration | FunctionExpression>node));
+                    if (!returnType || !returnType.symbol) {
+                        break;
+                    }
+
+                    // collect all the return statements
+                    // check that a property access expression exists in there and that it is a handler
+                    const retStmts: ReturnStatement[] = [];
+                    getReturnStmts(node, retStmts);
+
+                    if (isPromiseType(returnType) && retStmts.length > 0) {
+                        diags.push(createDiagnosticForNode(isVariableDeclaration(node.parent) ? node.parent.name : node, Diagnostics.This_may_be_converted_to_use_async_and_await));

```

Comment:
how can this be a `variableDeclaration`? we are sure the node is a returnStatement.

Reply:
Also. these nodes can span multiple lines. i would not put the span under the whole node, this way you would get teh suggestion deep inside a the handler in the `then` method. i would put the span on the `p.then` / `p.catch` / `p.finally` and not the whole expression.

Reply:
never mind my previous comments. i thought we are putting the span on the return statements.. now i see you are putting it on the function node. 

I think we should be putting the span on the `then`/`catch`/`finally` and not the function.

Reply:
I think the span should be on the entire function as we need to put the async keyword before the signature. Or should the two transformations: (1) adding async keyword (2) refactoring ` then / catch / finally ` be two different suggestions?